File: knowledge_inferer/conflict_resolver.py
"""
Conflict Resolver - Giải quyết xung đột khi nhiều rules cùng match
Implement 3 strategies cơ bản: Refractoriness, Specificity, Priority
"""

from typing import List, Dict, Any, Set, Tuple
from enum import Enum


class ConflictStrategy(Enum):
    """Các chiến lược giải quyết xung đột"""
    REFRACTORINESS = "refractoriness"  # Không kích hoạt rule đã dùng với cùng facts
    SPECIFICITY = "specificity"        # Ưu tiên rules có nhiều conditions hơn
    PRIORITY = "priority"              # Ưu tiên theo severity/priority
    RECENCY = "recency"                # Ưu tiên facts mới nhất


class ConflictResolver:
    """
    Conflict Resolver - Giải quyết xung đột khi nhiều rules cùng được kích hoạt
    
    Trong inference engine, khi nhiều rules cùng match với working memory,
    cần có chiến lược để quyết định rule nào được kích hoạt trước.
    """
    
    def __init__(self, strategies: List[ConflictStrategy] = None):
        """
        Args:
            strategies: Danh sách strategies theo thứ tự ưu tiên
                       Mặc định: [REFRACTORINESS, PRIORITY, SPECIFICITY]
        """
        if strategies is None:
            strategies = [
                ConflictStrategy.REFRACTORINESS,
                ConflictStrategy.PRIORITY,
                ConflictStrategy.SPECIFICITY
            ]
        self.strategies = strategies
        self.fired_rules: Set[Tuple[str, frozenset]] = set()
    
    def resolve(self, 
                candidate_rules: List[Dict[str, Any]], 
                working_memory_snapshot: Dict[str, Any]) -> List[Dict[str, Any]]:
        """
        Giải quyết xung đột và trả về danh sách rules theo thứ tự ưu tiên
        
        Args:
            candidate_rules: Danh sách các rules có thể kích hoạt
            working_memory_snapshot: Snapshot của working memory hiện tại
            
        Returns:
            Danh sách rules đã được sắp xếp theo thứ tự ưu tiên
        """
        if not candidate_rules:
            return []
        
        filtered_rules = candidate_rules.copy()
        
        for strategy in self.strategies:
            if strategy == ConflictStrategy.REFRACTORINESS:
                filtered_rules = self._apply_refractoriness(filtered_rules, working_memory_snapshot)
            elif strategy == ConflictStrategy.SPECIFICITY:
                filtered_rules = self._apply_specificity(filtered_rules)
            elif strategy == ConflictStrategy.PRIORITY:
                filtered_rules = self._apply_priority(filtered_rules)
            elif strategy == ConflictStrategy.RECENCY:
                filtered_rules = self._apply_recency(filtered_rules, working_memory_snapshot)
        
        return filtered_rules
    
    def _apply_refractoriness(self, 
                             rules: List[Dict[str, Any]], 
                             wm_snapshot: Dict[str, Any]) -> List[Dict[str, Any]]:
        """
        Strategy 1: Refractoriness
        Loại bỏ các rules đã được kích hoạt với cùng tập facts
        
        Nguyên tắc: Một rule không được kích hoạt lại với cùng một tập facts
        để tránh vòng lặp vô hạn
        """
        filtered = []
        
        for rule in rules:
            rule_id = rule['rule_id']
            
            matched_fact_ids = frozenset(rule.get('matched_facts', []))
            
            rule_signature = (rule_id, matched_fact_ids)
            
            if rule_signature not in self.fired_rules:
                filtered.append(rule)
        
        return filtered
    
    def _apply_specificity(self, rules: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """
        Strategy 2: Specificity
        Ưu tiên các rules có nhiều conditions hơn (cụ thể hơn)
        
        Nguyên tắc: Rule càng cụ thể (nhiều điều kiện) thì càng ưu tiên
        """
        if not rules:
            return rules
        
        sorted_rules = sorted(
            rules,
            key=lambda r: len(r.get('conditions', [])),
            reverse=True
        )
        
        return sorted_rules
    
    def _apply_priority(self, rules: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """
        Strategy 3: Priority
        Ưu tiên theo severity/priority của rule
        
        Nguyên tắc: high > medium > low
        """
        severity_order = {
            'high': 3,
            'medium': 2,
            'low': 1,
            None: 0
        }
        
        sorted_rules = sorted(
            rules,
            key=lambda r: severity_order.get(
                r.get('result', {}).get('severity', None), 0
            ),
            reverse=True
        )
        
        return sorted_rules
    
    def _apply_recency(self, 
                      rules: List[Dict[str, Any]], 
                      wm_snapshot: Dict[str, Any]) -> List[Dict[str, Any]]:
        """
        Strategy 4: Recency
        Ưu tiên rules match với facts mới nhất
        
        Nguyên tắc: Facts càng mới thì càng ưu tiên
        """
        return rules
    
    def mark_rule_fired(self, rule_id: str, matched_facts: List[str]):
        """
        Đánh dấu một rule đã được kích hoạt với tập facts cụ thể
        
        Args:
            rule_id: ID của rule
            matched_facts: Danh sách fact_ids mà rule này match
        """
        rule_signature = (rule_id, frozenset(matched_facts))
        self.fired_rules.add(rule_signature)
    
    def reset(self):
        """Reset lịch sử các rules đã kích hoạt"""
        self.fired_rules.clear()
    
    def get_fired_rules_count(self) -> int:
        """Lấy số lượng rules đã kích hoạt"""
        return len(self.fired_rules)
    
    def select_next_rule(self, 
                        candidate_rules: List[Dict[str, Any]], 
                        working_memory_snapshot: Dict[str, Any]) -> Dict[str, Any]:
        """
        Chọn rule tiếp theo để kích hoạt
        
        Args:
            candidate_rules: Danh sách các rules có thể kích hoạt
            working_memory_snapshot: Snapshot của working memory
            
        Returns:
            Rule được chọn để kích hoạt (hoặc None nếu không có)
        """
        resolved = self.resolve(candidate_rules, working_memory_snapshot)
        
        if not resolved:
            return None
        
        return resolved[0]
    
    def explain_selection(self, 
                         selected_rule: Dict[str, Any], 
                         all_candidates: List[Dict[str, Any]]) -> str:
        """
        Giải thích tại sao rule này được chọn
        
        Returns:
            Chuỗi giải thích
        """
        if not selected_rule:
            return "Không có rule nào được chọn."
        
        explanation_parts = [
            f"Rule {selected_rule['rule_id']} được chọn vì:"
        ]
        
        num_conditions = len(selected_rule.get('conditions', []))
        explanation_parts.append(f"- Có {num_conditions} điều kiện (specificity)")
        
        severity = selected_rule.get('result', {}).get('severity', 'unknown')
        explanation_parts.append(f"- Mức độ nghiêm trọng: {severity} (priority)")
        
        matched_facts = selected_rule.get('matched_facts', [])
        rule_signature = (selected_rule['rule_id'], frozenset(matched_facts))
        if rule_signature not in self.fired_rules:
            explanation_parts.append("- Chưa được kích hoạt với tập facts này (refractoriness)")
        
        other_rules = [r for r in all_candidates if r['rule_id'] != selected_rule['rule_id']]
        if other_rules:
            explanation_parts.append(f"- Ưu tiên hơn {len(other_rules)} rules khác")
        
        return "\n".join(explanation_parts)
    
    def __repr__(self):
        return f"ConflictResolver(strategies={[s.value for s in self.strategies]}, fired={len(self.fired_rules)})"

File: src/core/ConflictResolver.js
/**
 * Conflict Resolver - Giải quyết xung đột khi nhiều rules cùng match
 * Implement 3 strategies: Refractoriness, Specificity, Priority
 */

export const ConflictStrategy = {
  REFRACTORINESS: 'refractoriness',
  SPECIFICITY: 'specificity',
  PRIORITY: 'priority',
  RECENCY: 'recency'
};

export class ConflictResolver {
  constructor(strategies = null) {
    this.strategies = strategies || [
      ConflictStrategy.REFRACTORINESS,
      ConflictStrategy.PRIORITY,
      ConflictStrategy.SPECIFICITY
    ];
    this.firedRules = new Set(); // Set of "ruleId:factIds" signatures
  }

  /**
   * Giải quyết xung đột và trả về danh sách rules theo thứ tự ưu tiên
   * @param {Array} candidateRules - Danh sách rules có thể kích hoạt
   * @param {Object} workingMemorySnapshot - Snapshot của working memory
   * @returns {Array} Rules đã được sắp xếp
   */
  resolve(candidateRules, workingMemorySnapshot) {
    if (!candidateRules || candidateRules.length === 0) {
      return [];
    }

    let filteredRules = [...candidateRules];

    for (const strategy of this.strategies) {
      if (strategy === ConflictStrategy.REFRACTORINESS) {
        filteredRules = this._applyRefractoriness(filteredRules);
      } else if (strategy === ConflictStrategy.SPECIFICITY) {
        filteredRules = this._applySpecificity(filteredRules);
      } else if (strategy === ConflictStrategy.PRIORITY) {
        filteredRules = this._applyPriority(filteredRules);
      } else if (strategy === ConflictStrategy.RECENCY) {
        filteredRules = this._applyRecency(filteredRules, workingMemorySnapshot);
      }
    }

    return filteredRules;
  }

  /**
   * Strategy 1: Refractoriness
   * Loại bỏ các rules đã được kích hoạt với cùng tập facts
   */
  _applyRefractoriness(rules) {
    const filtered = [];

    for (const rule of rules) {
      const ruleId = rule.rule_id;
      const matchedFacts = rule.matched_facts || [];
      const signature = this._createSignature(ruleId, matchedFacts);

      if (!this.firedRules.has(signature)) {
        filtered.push(rule);
      }
    }

    return filtered;
  }

  /**
   * Strategy 2: Specificity
   * Ưu tiên các rules có nhiều conditions hơn (cụ thể hơn)
   */
  _applySpecificity(rules) {
    return rules.sort((a, b) => {
      const aConditions = a.conditions?.length || 0;
      const bConditions = b.conditions?.length || 0;
      return bConditions - aConditions; // Descending
    });
  }

  /**
   * Strategy 3: Priority
   * Ưu tiên theo severity/priority của rule
   */
  _applyPriority(rules) {
    const severityOrder = {
      'high': 3,
      'medium': 2,
      'low': 1
    };

    return rules.sort((a, b) => {
      const aSeverity = a.result?.severity || a.severity;
      const bSeverity = b.result?.severity || b.severity;
      const aOrder = severityOrder[aSeverity] || 0;
      const bOrder = severityOrder[bSeverity] || 0;
      return bOrder - aOrder; // Descending
    });
  }

  /**
   * Strategy 4: Recency (placeholder)
   */
  _applyRecency(rules, workingMemorySnapshot) {
    return rules;
  }

  /**
   * Tạo signature duy nhất cho rule + facts
   */
  _createSignature(ruleId, matchedFacts) {
    const sortedFacts = [...matchedFacts].sort();
    return `${ruleId}:${sortedFacts.join(',')}`;
  }

  /**
   * Đánh dấu một rule đã được kích hoạt
   */
  markRuleFired(ruleId, matchedFacts) {
    const signature = this._createSignature(ruleId, matchedFacts);
    this.firedRules.add(signature);
  }

  /**
   * Reset lịch sử các rules đã kích hoạt
   */
  reset() {
    this.firedRules.clear();
  }

  /**
   * Lấy số lượng rules đã kích hoạt
   */
  getFiredRulesCount() {
    return this.firedRules.size;
  }

  /**
   * Chọn rule tiếp theo để kích hoạt
   */
  selectNextRule(candidateRules, workingMemorySnapshot) {
    const resolved = this.resolve(candidateRules, workingMemorySnapshot);
    return resolved.length > 0 ? resolved[0] : null;
  }

  /**
   * Giải thích tại sao rule này được chọn
   */
  explainSelection(selectedRule, allCandidates) {
    if (!selectedRule) {
      return 'Không có rule nào được chọn.';
    }

    const parts = [`Rule ${selectedRule.rule_id} được chọn vì:`];

    const numConditions = selectedRule.conditions?.length || 0;
    parts.push(`- Có ${numConditions} điều kiện (specificity)`);

    const severity = selectedRule.result?.severity || selectedRule.severity || 'unknown';
    parts.push(`- Mức độ nghiêm trọng: ${severity} (priority)`);

    const matchedFacts = selectedRule.matched_facts || [];
    const signature = this._createSignature(selectedRule.rule_id, matchedFacts);
    if (!this.firedRules.has(signature)) {
      parts.push('- Chưa được kích hoạt với tập facts này (refractoriness)');
    }

    const otherRules = allCandidates.filter(r => r.rule_id !== selectedRule.rule_id);
    if (otherRules.length > 0) {
      parts.push(`- Ưu tiên hơn ${otherRules.length} rules khác`);
    }

    return parts.join('\n');
  }

  toString() {
    return `ConflictResolver(strategies=${this.strategies.join(',')}, fired=${this.firedRules.size})`;
  }
}
